src/v/coproc/script_context.cc

```diff
@@ -103,9 +104,40 @@ ss::future<> script_context::do_execute() {
     });
 }
 
+void script_context::notify_waiters() {
+    auto updates = std::exchange(_updates, {});
+    for (auto& [ntp, update] : updates) {
+        auto found = _routes.find(update.source);
+        if (found != _routes.end()) {
+            found->second->wctx.offsets.erase(ntp);
+        }
+        for (auto& p : update.ps) {
+            p.set_value();
+        }
+    }
+}
+
+ss::future<> script_context::remove_output(
+  const model::ntp& source, const model::ntp& materialized) {
+    auto found = _updates.find(materialized);
+    if (found == _updates.end()) {
+        found = _updates.emplace(materialized, update{.source = source}).first;
+    }
+    vassert(found->second.source == source, "Mismatched");
+    found->second.ps.emplace_back();
+    return found->second.ps.back().get_future();
+}
+
 ss::future<> script_context::shutdown() {
     _abort_source.request_abort();
-    return _gate.close();
+    co_await _gate.close();
+    auto updates = std::exchange(_updates, {});
+    for (auto& [ntp, update] : updates) {
+        for (auto& p : update.ps) {
+            p.set_exception(wait_future_stranded(
```

Comment:
would any fibers waiting on a promise here be inside the gate being closed above?

Reply:
By fibers do you mean a coproc fiber or like a thread of execution? I think the idea here was to cancel pending updates after the gate is closed and other threads of execution may operate independently of that.

Reply:
what I'm doing a bad job of asking is: if a fiber were waiting on the exception to be set to make progress, and that fiber were inside the gate closed a few lines above, then would a deadlock occur?

Reply:
ah yes it would but the fibers never wait on this condition, only consumers of this public interface would ever use this notification mechanism.